cilium: refactoring and improvements for lb

borkmann opened the pull request:
(see commit desc)

borkmann committed:
cilium: Enable load-balancer mode to connect to kube-apiserver

Allow connecting to the kube-apiserver in lb-only datapath mode. This allows
to utilize the K8s control plane for orchestration as in CNI mode and enables
persistent storage of services and backends.

Note that in this mode having local Pods in the cluster is currently not
supported and therefore the BPF code generation skipped in endpoints package.

Co-developed-by: Aditi Ghag <aditi@cilium.io>
Signed-off-by: Aditi Ghag <aditi@cilium.io>
Co-developed-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Add service infrastructure for per-service health

This adds new infrastructure for building custom health checkers for services.
The current backend health infrastructure has the limitation in that it is
global to the backend. This is fine if a backend is generally taken out of
service for maintenance updates, however, this approach will not work if a
L4 backend contains multiple L7 services. In this case the backend health
property/state needs to be more fine-grained and tied to the specific service
instead of being global.

We achieve this by:

1) Changing UpdateBackendsState() to call into UpdateBackendsStateMultiple().
   The latter is reworked into a reusable function, taking a custom service
   map to update and a boolean whether to also quarantine the whole backend.
   The per-service quarantining is called via UpdateBackendStateServiceOnly().

2) The Backend Go object is not shared between services, so state transition
   can be updated via UpdateBackendStateServiceOnly() and not affecting other
   services. The latter skips updating the actual backend state. So the per-
   service state can be quarantined but the global state still active.

3) The BPF datapath in this case still "takes out" the backend from the BPF
   service map, so that new connections do not pick the given backend to
   redirect traffic to. It does so by reducing the struct lb{4,6}_service
   count field. Also, we reuse the pad[2] fields to explicitly add the
   qcount field to denote a quarantine count. These backends are present
   in the service map via struct lb{4,6}_key with a backend_slot lookup range
   of [svc->count, svc->count + svc->qcount).

4) Restoration takes the BPF map information into account in order to rebuild
   the internal per-service quarantine representation before a sync with the
   kube-apiserver happens. DumpServiceMaps() and svcBackend() set the per-
   service backend state.

4) This work also adds a registration mechanism for building custom health checkers
   via HealthChecker interface with hooks for UpsertService() and corresponding
   DeleteService() as well as a mechanism to register callbacks via SetCallback().
   The HealthCheckCallback() API can then be used to update a per-service
   backend state. See also the TestHealthCheckCB() unit test in the subsequent
   patch.

5) The Service object also plumbs through annotations from the Kubernetes service.
   This helps for the health checkers to plumb through configuration data when
   registering the service at the health checker plugin.

Co-developed-by: Aditi Ghag <aditi@cilium.io>
Signed-off-by: Aditi Ghag <aditi@cilium.io>
Co-developed-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, test: Extend service tests with health check callbacks

Extend the service tests to make use of the health checker callback infra
to simulate that when health check for one frontend fails, we don't take
down the whole backend through this facility but just the backend corresponding
to the given service.

Co-developed-by: Aditi Ghag <aditi@cilium.io>
Signed-off-by: Aditi Ghag <aditi@cilium.io>
Co-developed-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Add bpf-lb-external-control-plane switch for old-style docker mode

Docker style mode is still in use, therefore add a new --bpf-lb-external-control-plane
switch (default is false) where users can opt into the old behavior when it is being
deployed in non-k8s environments. Adopt the CI tests accordingly by adding
--bpf-lb-external-control-plane=true to their cmdline.

"bpf-lb-external-control-plane" because it needs to be provided externally since the
only thing that the agent listens on is the swagger API for installing frontends and
backends.

Co-developed-by: Aditi Ghag <aditi@cilium.io>
Signed-off-by: Aditi Ghag <aditi@cilium.io>
Co-developed-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

mhofstetter commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -49,6 +49,7 @@ cilium-agent [flags]
       --bpf-lb-maglev-table-size uint                             Maglev per service backend table size (parameter M) (default 16381)
       --bpf-lb-map-max int                                        Maximum number of entries in Cilium BPF lbmap (default 65536)
       --bpf-lb-mode string                                        BPF load balancing mode ("snat", "dsr", "hybrid") (default "snat")
+      --bpf-lb-no-control-plane                                   BPF load balancer uses an externally-provided control plane
```
-> rename to `--bpf-lb-disable-k8s-controlplane` / `--bpf-lb-disable-k8s-cp`

borkmann replied:
or wdyt of `bpf-lb-external-control-plane`?

mhofstetter replied:
sounds good to me

mhofstetter commented on pkg/option/config_test.go:
```diff
@@ -1382,7 +1382,7 @@ func TestDaemonConfig_StoreInFile(t *testing.T) {
   	RestoreState: false,
   	KeepConfig:   false,
-  	... // 5 ignored and 312 identical fields
+  	... // 5 ignored and 313 identical fields
```
squashing this commit with the one that is actually introducing the new field?

borkmann replied:
yeap will do

qmonnet reviewed (approved):
Looks good to me, nice work!

mhofstetter commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -49,6 +49,7 @@ cilium-agent [flags]
       --bpf-lb-maglev-table-size uint                             Maglev per service backend table size (parameter M) (default 16381)
       --bpf-lb-map-max int                                        Maximum number of entries in Cilium BPF lbmap (default 65536)
       --bpf-lb-mode string                                        BPF load balancing mode ("snat", "dsr", "hybrid") (default "snat")
+      --bpf-lb-no-control-plane                                   BPF load balancer uses an externally-provided control plane
```
IMO it would be nicer if we could place this as part of the Hive Cell providing the K8sClient (and therefore ~ the K8s Controlplane integration) itself (https://github.com/cilium/cilium/blob/main/pkg/k8s/client/cell.go#L55) and naming it accordingly.

All the places that currently would check for the new flag could check whether the K8sClient is enabled or not.

But this change would only be possible if the K8sClient is available in all the places that check for the controlplane.

borkmann replied:
Okay, sounds good, I will look into this as a follow-up.

mhofstetter commented on daemon/cmd/api_handlers.go:
```diff
@@ -106,56 +106,61 @@ func ciliumAPIHandlers(dp promise.Promise[*Daemon], cfg *option.DaemonConfig, _
+	out.ServicePutServiceIDHandler = wrapAPIHandler(dp, putServiceIDHandler)
+
+	// If no control plane is set, we only expose service API to the user for programming.
+	if option.Config.LoadBalancerNoControlPlane {
```
> we only expose service API to the user for programming

btw: this function only adds the api handlers that are still implemented "in the daemon" there are also other api handlers that are already provided by the respective cells directly.

As more and more of the api handlers will be migrated into the respective cells it would be actually nice to keep the list of api handlers that don't work if the K8s CP isn't available.

borkmann replied:
Ok, fixed by enabling all + added context to why this is ok. Thanks for pointing this out!

borkmann committed:
cilium, api-handler: Do not exclude APIs in lb-only mode

Just install all API handlers in lb-ony mode. The reason back then to disable API
parts was due to b052272d550a ("daemon: Disable parts of Cilium API in LB mode")
which was for a case where the Cilium CNI plugin was installed, a K8s cluster
still had some endpoints and then the agent in lb-only mode tried to restore
them but with errors. This is fine since Docker-only mode is targeted for non-
K8s anyway, and also two Cilium instances in hostns (CNI + lb-only) would step
on each other and is not supported at the moment.

Also, Marco added that this daemon function only adds the API handlers that are
still implemented "in the daemon", but there are also other API handlers that
are already provided by the respective cells directly. More of the latter will
be migrated in future.

Leaving this here as separate commit for context.

Suggested-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

mhofstetter reviewed (approved):
LGTM - thanks for addressing my inputs! :pray:

aspsk reviewed (approved):
Cool, thanks! :shipit:
